dask/array/core.py

```diff
@@ -1430,7 +1430,7 @@ def to_hdf5(self, filename, datapath, **kwargs):
         """
         return to_hdf5(filename, datapath, self, **kwargs)
 
-    def to_dask_dataframe(self, columns=None, index=None):
+    def to_dask_dataframe(self, columns=None, index=None, meta_df=None):
```

Comment:
I don't think that this name immediately explains the concept it's capturing. Can you think of alternatives?

Perhaps `dataframe_class`? Or does that not convey that this is the concrete dataframe that make up the partitions?

You'll need to add this parameter to the docstring.

Reply:
Does this need to be instantiated, or just the class?

Reply:
I don't find `target_dataframe` a meaningful kwarg name - my first thought from that name was this was the class of the chunk type.

Since this is an instantiated object that will be used as the metadata for the target, something with `meta` would be more consistent with the rest of the api. I'd probably go with `meta` here (I realize this contradicts Tom's suggestion above, sorry Tom :/). Since it's an instantiated object used as metadata it seems fitting to keep the existing verbage - many of our other conversion routines to dataframes also have this kwarg.

Reply:
@TomAugspurger are you also okay with `meta`?
Or one more option that comes to my mind is `target_meta`/`meta_target`?

Reply:
My only hesitation with using `meta` is that it overloads (or expands) the definition of meta a bit. Right now it's used to identify the structure of the dataset: columns, index(es), and their dtypes. If we also allow it to mean the concrete container type, then what does that imply. If I have something like

```python
df = dd.from_pandas(pd.DataFrame({"A": [1, 2, 3]}), 2)
df.map_partitions(lambda x: x + 1, meta=cudf.DataFrame({"A": int}))`
```

would you expect that to copy from host to device memory by calling `cudf.DataFrame._constructor` on the result of the `lambda`? According to `meta`, the concrete type should be a `cudf.DataFrame`. But the function doesn't do anything with host -> device, so it would need to be done implicitly.

And to be clear: I'm not against defining that behavior, but we should do it intentionally.

Reply:
> If we also allow it to mean the concrete container type, then what does that imply?

Don't we already define that behavior? If I'm using `cudf` with dask dataframe, isn't `meta` a cudf object?

> would you expect that to copy from host to device memory by calling cudf.DataFrame._constructor on the result of the lambda?

Currently we trust the user that `meta` is correct, and sometimes warn/error if that isn't true. In this case we'd trust the user that the output is a `cudf` type, but wouldn't do the conversion ourselves. I believe this matches current conventions, and is the least-surprising choice. `meta` is there to hint to dask-dataframe the structure of the underlying dataframe when inference may not be able to handle it - it's a type hint, not a type conversion.

Reply:
> Don't we already define that behavior? If I'm using cudf with dask dataframe, isn't meta a cudf object?

I'm not sure.

Good point about us trusting `meta`. That's consistent with not implicitly calling `meta._constructor` on the output, since we're trusting that it's correct.

So I'm probably fine with using `meta` here.

Reply:
Okay, updated the param name to be `meta`.